tests/library/mpi/redistribute_test.py

```diff
@@ -0,0 +1,268 @@
+    size = commworld.Get_size()
+    even_size = (size // 2) * 2
+
+    if size < 2:
+        raise ValueError("Please run this test with at least two processes.")
+
+    if rank == 0:
+        sdfg = matrix_2d_2d_3.to_sdfg()
+        func = sdfg.compile()
+    commworld.Barrier()
+    if rank > 0:
+        sdfg = dace.SDFG.from_file(".dacecache/{n}/program.sdfg".format(n=matrix_2d_2d_3.name))
+        func = CompiledSDFG(sdfg, ReloadableDLL(".dacecache/{n}/build/lib{n}.so".format(n=sdfg.name), sdfg.name))
+    commworld.Barrier()
+
+    A = np.arange(64 * even_size * even_size, dtype=np.int32).reshape(8 * even_size, 8 * even_size)
+    lA = A.reshape(2, 4 * even_size, even_size // 2, 16).transpose(0, 2, 1, 3)
+    lB = A.reshape(2, 4 * even_size, even_size // 2, 16).transpose(2, 0, 1, 3)
+    if rank < even_size:
+        B = func(A=lA[rank // (even_size // 2), rank % (even_size // 2)].copy(), P=even_size)
+    else:
+        B = func(A=np.zeros((1, ), dtype=np.int32), P=even_size)
+
+    if rank < even_size:
+        assert (np.array_equal(B, lB[rank // 2, rank % 2]))
+
+
+@pytest.mark.mpi
+def test_redistribute_vector_2d_2d():
+    """
+    The numbers are example tile IDs, NOT MPI ranks. "(r)" means that the tile is a replica.
+     ____________________        _______________________        ___________
+    |____________________|  ->  |0____|1____|2____|3____|  ->  |0  __|zero_|
+                                |0(r)_|1(r)_|2(r)_|3(r)_|      |1____|zero_|
+                                                               |2____|zero_|
+                                                               |3____|zero_|
+    """
+
+    P = dace.symbol('P', dace.int32)
+
+    @dace.program
+    def vector_2d_2d(A: dace.int32[8 * P]):
+
+        a_grid = dace.comm.Cart_create([2, P // 2])
+        a_scatter_grid = dace.comm.Cart_sub(a_grid, [False, True], exact_grid=0)
+        a_bcast_grid = dace.comm.Cart_sub(a_grid, [True, False])
+        b_grid = dace.comm.Cart_create([P // 2, 2])
+        b_scatter_grid = dace.comm.Cart_sub(b_grid, [True, False], exact_grid=0)
+        b_bcast_grid = dace.comm.Cart_sub(b_grid, [False, True])
+
+        lA = np.empty_like(A, shape=(16, ))
+        a_subarr = dace.comm.BlockScatter(A, lA, a_scatter_grid, a_bcast_grid)
+        lB = np.zeros_like(A, shape=(16, ))
+        b_subarr = dace.comm.Subarray((8 * P, ), lB, process_grid=b_scatter_grid)
+        redistr = dace.comm.Redistribute(lA, a_subarr, lB, b_subarr)
+
+        return lB
+
+    from mpi4py import MPI
+    commworld = MPI.COMM_WORLD
+    rank = commworld.Get_rank()
+    size = commworld.Get_size()
+    even_size = (size // 2) * 2
+
+    if size < 2:
+        raise ValueError("Please run this test with at least two processes.")
+
+    if rank == 0:
+        sdfg = vector_2d_2d.to_sdfg()
+        func = sdfg.compile()
+    commworld.Barrier()
+    if rank > 0:
+        sdfg = dace.SDFG.from_file(".dacecache/{n}/program.sdfg".format(n=vector_2d_2d.name))
```

Comment:
maybe you should make a test util that only compiles on rank 0 and loads it on the rest, i.e., replacing lines 241-248 here.